src/main/java/org/candlepin/model/ContentCurator.java

```diff
@@ -645,68 +645,76 @@ public Set<Content> getChildrenContentOfProductsByUuids(Collection<String> produ
+                    UNION
+                    SELECT CAST(pcmap.child_uuid AS CHAR(32)), pgraph.depth + 1
+                    FROM parent_child_map pcmap
+                    JOIN product_graph pgraph ON pgraph.uuid = pcmap.parent_uuid
+                    WHERE pgraph.depth < 10
+                )
+                SELECT content.uuid, content.namespace, content.content_id, content.type,
+                       content.label, content.name, content.vendor, content.content_url,
+                       content.required_tags, content.release_ver, content.gpg_url,
+                       content.metadata_expire, content.arches, pc.enabled
+                FROM product_graph pgraph
+                JOIN cp_product_contents pc ON pc.product_uuid = pgraph.uuid
+                JOIN cp_contents content ON content.uuid = pc.content_uuid
+            """;
+
+        List<Object[]> results = entityManager.createNativeQuery(sql)
             .setParameter("owner_id", ownerId)
             .setParameter("start_date", now)
             .setParameter("end_date", now)
-            .getResultList());
-
-        Query prodQuery = entityManager.createNativeQuery(sql);
-        int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 2);
+            .getResultList();
 
-        // Get all of the derived products and provided products for the products we've found, recursively
-        for (Set<String> lastBlock = productUuids; !lastBlock.isEmpty();) {
-            Set<String> children = new HashSet<>();
+        List<ProductContent> activeContent = new ArrayList<>();
 
-            for (List<String> block : this.partition(lastBlock, blockSize)) {
-                children.addAll(prodQuery.setParameter("product_uuids", block)
-                    .getResultList());
-            }
-
-            productUuids.addAll(children);
-            lastBlock = children;
-        }
-
-        // Use the product UUIDs to select all related enabled content
-        jpql = "SELECT DISTINCT pc.content, pc.enabled FROM ProductContent pc " +
-            "WHERE pc.product.uuid IN (:product_uuids)";
-
-        Query contentQuery = entityManager.createQuery(jpql);
-        HashMap<Content, Boolean> activeContent = new HashMap<>();
-
-        for (List<String> block : this.partition(productUuids)) {
-            contentQuery.setParameter("product_uuids", block)
-                .getResultList()
-                .forEach(col -> {
-                    Content content = (Content) ((Object[]) col)[0];
-                    Boolean enabled = (Boolean) ((Object[]) col)[1];
-                    activeContent.merge(content, enabled, (v1, v2) ->
-                        v1 != null && v1.booleanValue() ? v1 : v2);
-                });
+        for (Object[] result : results) {
+            activeContent.add(productContentMapping(result));
         }
 
         return activeContent;
     }
 
+    protected ProductContent productContentMapping(Object[] result) {
```

Comment:
This has two problems:
1) the proper functionality of this method is very tightly coupled to the correctness of the array passed in -- which itself is entirely unchecked throughout the method -- and that is really only viable in the context of the query in question.

2) while we're building most of the content correctly, this doesn't include the required or "modified" product IDs. Additionally, because it's a concrete Content instance and not a Hibernate proxy, it will return an empty list instead of fetching them; which could have consequences should this method be used in the future for other functionality (e.g. the proposed "active content" endpoint).

This logic would be better served as a closure within the method executing the query so that it could be applied as a stream, and logically separated from the core functionality of the method, without being decoupled and at risk of being used elsewhere.

For addressing the second point, it might be best or easiest to simply fetch the content IDs, and then do a second lookup to get Hibernate instances of them. Yes, it's less efficient overall, but it saves us the potential pain of these various edge cases where manually building the object doesn't work completely for one reason or another.

If there are alternate means of obtaining Hibernate proxies which can be used to fully hydrate objects as needed, I'm also open to exploring those options.

Reply:
so I took the route of another query.
Done